package/MDAnalysis/core/selection.py

```diff
@@ -518,11 +518,19 @@ def __init__(self, parser, tokens):
         self.values = vals
 
     def apply(self, group):
-        mask = np.zeros(len(group), dtype=np.bool)
-        for val in self.values:
-            values = getattr(group, self.field)
-            mask |= [fnmatch.fnmatch(x, val) for x in values]
-        return group[mask].unique
+        # rather than work on group.names, cheat and look at the lookup table
+        nmattr = getattr(group.universe._topology, self.field)
+
+        matches = []  # list of passing indices
+        # iterate through set of known atom names, check which pass
+        for nm, ix in nmattr.namedict.items():
+            if any(fnmatch.fnmatch(nm, val) for val in self.values):
```

Comment:
So just out of curiosity I looked into how much `fnmatch` itself was a culprit in the slowdown we see on `develop`.

It looks like we can get a decent speedup by just switching over to regex by doing something along the lines of `fnmatch.translate(val)` combined with `reobj = re.compile` and `reobj .match()`.

Modifying the `develop` code to switch from pure fnmatch to regex gives a ~ 5x speed up, so definitely the approach used here is worth it. However I wonder if we can somehow leverage the improved performance of `re.match()` somewhere here?

I've tried:

```python
     def apply(self, group):
         # rather than work on group.names, cheat and look at the lookup table
         nmattr = getattr(group.universe._topology, self.field)
         vals = [ re.compile(fnmatch.translate(i)) for i in self.values ]

         matches = []  # list of passing indices
         # iterate through set of known atom names, check which pass
         for nm, ix in nmattr.namedict.items():
             if any(bool(val.match(nm)) for val in vals):
                 matches.append(ix)
```

But the performance difference is at best within error. It might be that we could do a better job by making `grab_not_keyword` return regex objects instead?

Reply:
iirc `re` already caches things so the speedup from explicitly `re.compile` before use is minimal.

Something I haven't looked at yet is here we're matching against all possible names in the Universe, but really we only need to match all names in the selection scope.  So you could first create a set of `nmidx[group.ix]` and only run fnmatch/re against that subset.

Reply:
So for context this is what I did on develop to get the speed up (+/- indicating the pre and post regex code):

```python
     @return_empty_on_apply
     def apply(self, group):
         mask = np.zeros(len(group), dtype=np.bool)
+        values = getattr(group, self.field)
         for val in self.values:
-            values = getattr(group, self.field)
-            mask |= [fnmatch.fnmatch(x, val) for x in values]
+            reobj = re.compile(fnmatch.translate(val))
+            mask |= [bool(reobj.match(x)) for x in values]
         return group[mask].unique
```

Pre-regex my million atom benchmark got:
alpha carbon selection: 1.32 s
water oxygen selection: 2.72 s

Post-regex, my million atom benchmark got:
alpha carbon selection: 269 ms
water oxygen selection: 712 ms

It's worth noting that the cost of creating the regex object is probably high, so it's possible that on smaller systems the gains aren't actually real.